Decorator versioning

Firtzberg opened the pull request:
Based on yersterday's discussion this is the proposed structure for the decorators.
The linguist-generated=true in .gitattributes seems not to have effect. Any suggestions?

Firtzberg committed:
Extracted CreshThresholdDecorator into own versioned component.

Firtzberg committed:
Excluding fabricated files from diff.
Preventing export of fabrication and linting configurations.

jpmarcotte commented:
> The linguist-generated=true in .gitattributes seems not to have effect. Any suggestions?

IIRC, that sometimes took a bit to take effect, potentially only after it got merged to master? It's been a while since I've used it.

jpmarcotte commented on CrashedThresholdWorkerDecoratorV1/fab/CrashedThresholdDecorator/AwareTrait.php:
```diff
@@ -0,0 +1,46 @@
+<?php
```
I think from a repo cleanliness standpoint, putting all of these into a single `Decorators/` or `WorkerDecorators/` directory instead of into the root would probably be a good idea.

Firtzberg committed:
Created folder for all WorkerDecorator components and moved CrashedThresholdDecorator into it.

Firtzberg committed:
Moved Worker namespace to Worker1.

Firtzberg commented:
Updated. I've also changed the position of the basic classes (WorkerInterface, DecoratorInterface) to WorkerV1.
This is easiest to understand by looking at the [composer.json](https://github.com/neighborhoods/KojoWorkerDecoratorComponent/pull/14/files#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34).

jpmarcotte commented on composer.json:
```diff
@@ -42,6 +42,14 @@
+        "WorkerV1\\src",
+        "WorkerV1\\fab"
+      ],
+      "Neighborhoods\\KojoWorkerDecoratorComponent\\WorkerDecorators\\CrashedThresholdV1\\": [
+        "WorkerDecorators\\CrashedThresholdV1\\src",
+        "WorkerDecorators\\CrashedThresholdV1\\fab"
```
I think my concern with this pattern is that composer.json and the buphalo script both have to be updated for every decorator added. I wonder if a structure like this might not make things more accessible / easier to contribute / less prone to mistakes: i.e. Something that is more along the lines of our Pit of Success value (easy to do the right thing and harder to do the wrong thing).

```suggestion
      "Neighborhoods\\KojoWorkerDecoratorComponent\\WorkerDecorators\\": [
        "WorkerDecorators\\src",
        "WorkerDecorators\\fab"
```

Firtzberg replied:
That would make it harder for users to add the needed paths if they want specific decorators only.
With the current approach it's
```
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerV1')
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerDecorators/CrashedThresholdV1')
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerDecorators/ExceptionHandlingV1')
```

With the proposed change it would be
```
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerV1')
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerDecorators/fab/CrashedThresholdV1')
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerDecorators/src/CrashedThresholdV1')
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerDecorators/fab/ExceptionHandlingV1')
->addSourcePath('vendor/neighborhoods/kojo-worker-decorator-component/WorkerDecorators/src/ExceptionHandlingV1')
```

jpmarcotte replied:
Yeah, it's definitely a tradeoff, but adding `src/<Component>` and `fab/<Component>` is already part of the standard process for every component you want to add to a container. And really, thinking about this more makes me feel like it might be easiest (and matches other existing patterns that we already use) to have the file structure look like this:

```
src/
    WorkerV1/
    WorkerDecorators/
fab/
    WorkerV1/
    WorkerDecorators/
```

(I'm probably creating a ListingService specific wrapper for the container builder, so I've been thinking about this kind of thing for a while, and maintaining the existing pattern feels like more of a Pit of Success than trying to come up with something clever here.)

jpmarcotte replied:
(And yes, I realize I was the one that came up with the original idea of putting separate `src/` and `fab/` directories in each one, instead of vice versa, but after seeing it in practice, I like it less and less)

Firtzberg replied:
No problem.
Another difference in addition to the structure are the method names in the generated AwareTraits. When having the src folder in the root the method names are longest. With the approach I initially used they were super short.

jpmarcotte commented on bin/buphalo:
```diff
@@ -3,10 +3,26 @@
+vendor/bin/buphalo
+
+# shellcheck disable=SC2034
+Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__SourceDirectoryPath=$PWD/WorkerDecorators/CrashedThresholdV1/src \
+Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__FabricationDirectoryPath=$PWD/WorkerDecorators/CrashedThresholdV1/fab \
+Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__NamespacePrefix=Neighborhoods\\KojoWorkerDecoratorComponent\\WorkerDecorators\\CrashedThresholdV1\\ \
```
Related to the other comment. Would also require changing the file structure a bit.
```suggestion
Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__SourceDirectoryPath=$PWD/WorkerDecorators/src \
Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__FabricationDirectoryPath=$PWD/WorkerDecorators/fab \
Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__NamespacePrefix=Neighborhoods\\KojoWorkerDecoratorComponent\\WorkerDecorators\\ \
```

Firtzberg replied:
How about I change the script to loop over all subfolders inside WorkerDecorators?

Firtzberg replied:
I've modified it to loop over all subfolders in WorkerDecorators

jpmarcotte reviewed (commented):
Additionally, if you're changing the worker to `WorkerV1`, it's probably worth naming the decorators directory/namespace `WorkerDecoratorsV1` or `WorkerV1Decorators`, since a breaking change to Worker pattern would probably necessitate creating an entirely new set of Decorators.

Firtzberg committed:
Versioning Decorators folder based on corresponding Worker version.

Firtzberg committed:
Added gitattributes to Worker1.

Firtzberg committed:
Fixed buphalo file and improved exclusion of fab folders in linting tools.

Firtzberg committed:
Buphaloing all decorators without listing them.

Firtzberg committed:
Renamed WorkerDecoratorsV1 to WorkerV1Decorators.

Firtzberg committed:
Moved RetryThresholdDecorator to its versioned folder.

Firtzberg committed:
Moved ExceptionHandlingDecorator to its versioned folder.

Firtzberg committed:
Moved UserlandPdoDecorator to its versioned folder.

Firtzberg committed:
Moved ReschedulingDecorator to its versioned folder.

Firtzberg committed:
Removed unused WorkerAwareTrait.

Firtzberg committed:
Cleaned up src.

Firtzberg committed:
Single gitattributes files

Firtzberg committed:
Fixed file endings - added blank where missing

Firtzberg committed:
moved src and fab to project root.

Firtzberg committed:
Fixed Worker Builder template DI configuration. Added retry threshold decorator.

Firtzberg committed:
Finegrained kojo worker decorator paths in container template.

Firtzberg committed:
Fixed template for worker builder DI service.

Firtzberg committed:
Fix Worker Builder.

Firtzberg commented:
I think this is how the src folder will look like. I need to update the documentation.
Should I keep all the documentation in the README or split it up into individual components?

I personally think documenting possibly different versions of the same decorator in one place is confusing. There's also the possibility to have different worker versions.
I'll make some documentation changes as I see fit on a different branch and create a PR to this one when ready. In case you don't like it a can just abandon it

Firtzberg commented:
**Buphalo templates**

1. Currently we have a [template tree](https://github.com/neighborhoods/KojoWorkerDecoratorComponent/tree/SERCOMP-3/template-tree/V1/KojoWorkerDecoratorComponent) with 2 subfolders. One subfolder is for worker fabrication, while the other is for decorator fabrication. I propose to indicate the Worker version in the templates path.
I would like to rename `Worker` to `WorkerV1` and `Decorator` to `WorkerV1Decorator`.

2. You notice that the path already contains a `V1` segment. Correct me if I'm wrong. In my understanding that `V1` indicates that the templates are meant for buphalo version 1. I guess if there were a buphalo 2 release which is incompatible with `V1` templates, we would create a `V2` folder and adapt the templates.

3. You also notice that the path contains a `KojoWorkerDecoratorComponent` segment. This is meant to prevent template path collision with buphalo templates from other service agnostic components. This is good and no change is required.

4. Even after renaming `Worker` to `Worker1` and `Decorator` to `WorkerV1Decorator` the templates wouldn't be versioned. In case we want to make improvement to the templates, we would introduce possibly breaking changes. So inside `template-tree/V1/KojoWorkerDecoratorComponent/WorkerV1` we need subfolders indicating the version of the template.
We would end up with `template-tree/V1/KojoWorkerDecoratorComponent/WorkerV1/V1/PrimaryActorName.php`
Does this sound reasonable?

Firtzberg committed:
Renamed WorkerV1 to WorkerDecorationV1

Firtzberg committed:
Split documentation. Wrote untested upgrade guide.

jpmarcotte commented:
> You notice that the path already contains a V1 segment. Correct me if I'm wrong. In my understanding that V1 indicates that the templates are meant for buphalo version 1. I guess if there were a buphalo 2 release which is incompatible with V1 templates, we would create a V2 folder and adapt the templates.

Yeah, that's mostly correct. It may be that it's not from a Buphalo 2.0 _per se_, but a new version of the Tokenizer / Compiler rulesets that would require the templates to be keyed a little differently. Those could theoretically be released as something that's part of 1.X if there's still an option to use a customizable tokenization/compilation ruleset. Or it could be something entirely different—we haven't worked out what that looks like yet. That `v1` could also be for your own use if you decide to completely redo the architecture and use a whole different set of patterns (which is kind of related to point 4)...

> Even after renaming Worker to Worker1 and Decorator to WorkerV1Decorator the templates wouldn't be versioned. In case we want to make improvement to the templates, we would introduce possibly breaking changes. So inside template-tree/V1/KojoWorkerDecoratorComponent/WorkerV1 we need subfolders indicating the version of the template.
We would end up with template-tree/V1/KojoWorkerDecoratorComponent/WorkerV1/V1/PrimaryActorName.php
Does this sound reasonable?

Sounds reasonable. There's a handful of ways to do that, and it depends on how you want to manage it. I don't know if you necessarily need versioned templates, but assuming you do, you can
- continue with what you've got
- separate things into different trees and take advantage of the [multiple template trees feature](https://github.com/neighborhoods/Buphalo#using-multiple-template-trees). (Note that the multiple template trees feature is designed around looking for the specified path in any of the trees, so if you want to keep exclusive trees, you may want to consider multiple executions)
- Create new templates in the same tree by instead adding a new template, e.g. `PrimaryActorNameV2.php` (the generated Actor file path doesn't need to match the template file path.)
  - ListingService experimented for a bit by having different styles of repositories: one for when you want to `insert` and `update` separately, and one for when you wanted to `save` (`INSERT ... ON CONFLICT ... DO UPDATE ...`), so we'd have something like this:
    ```yml
      <PrimaryActorName>/Repository.php:
        template: PrimaryActorName/SaveRepository.php
        # or
        template: PrimaryActorName/InsertUpdateRepository.php
    ```
 
I think it depends on whether or not you think you'll want to update templates individually or update the entire tree.

Firtzberg committed:
Merge pull request #15 from neighborhoods/SERCOM-3-documentation

Decorator versioning with split documentation

arielallon commented:
FWIW, re `template-tree/V1/KojoWorkerDecoratorComponent/WorkerV1/V1/PrimaryActorName.php` I think it could help to clarify the various `V1` designations as to what they are are modifying, e.g. `template-tree/BuphaloV1/KojoWorkerDecoratorComponent/WorkerDecorationV1/TemplateV1/PrimaryActorName.php`

Firtzberg committed:
Restructured templates to be versioned

Firtzberg committed:
Upgraded to throwable diagnostic 4.

Firtzberg committed:
Minor improvements to readmes.

Firtzberg committed:
Simplified inclusion of throwable diagnostic components in worker container.

Firtzberg committed:
Merge branch 'master' into SERCOMP-3

Firtzberg committed:
Working upgrade guide.

Firtzberg committed:
Skipping container configuration when in cache.

jpmarcotte commented on bin/buphalo:
```diff
@@ -7,6 +7,6 @@ cd ..
 Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__FabricationDirectoryPath=$PWD/fab \
 Neighborhoods_Buphalo_V1_TargetApplication_BuilderInterface__NamespacePrefix=Neighborhoods\\KojoWorkerDecoratorComponent\\ \
-Neighborhoods_Buphalo_V1_TemplateTree_Map_Builder_FactoryInterface__TemplateTreeDirectoryPaths=default:$PWD/vendor/neighborhoods/buphalo/template-tree/V1,diagnostic:$PWD/vendor/neighborhoods/throwable-diagnostic-component/template-tree/BuphaloV1,worker:$PWD/template-tree/V1 \
+Neighborhoods_Buphalo_V1_TemplateTree_Map_Builder_FactoryInterface__TemplateTreeDirectoryPaths=default:$PWD/vendor/neighborhoods/buphalo/template-tree/V1,diagnostic:$PWD/vendor/neighborhoods/throwable-diagnostic-component/template-tree/BuphaloV1,kwdc:$PWD/template-tree/BuphaloV1 \
```
FYI: Buphalo makes no guarantees about its own templates. If you want to use them as a starting point, you should copy them to your own repo, but they are not part of our contract and changes to them may not be considered a breaking change.

jpmarcotte reviewed (approved):
Fairly quick look, things look good. It's a bit tough to say with certainty until I'm actually able to get my hands on it, though. Hopefully soon.
